Type punned dereference in libnml

A cast to int& is performed in libnml/nml/stat_msg.cc:38:
```c++
    cms->update((int&)(((RCS_STAT_MSG *) buf)->status));
```
This int& cast evokes a warning:
<code>libnml/nml/stat_msg.cc:38:48: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]</code>

It is a serious warning because the type of the class member 'status' (class RCS_STAT_MSG in libnml/nml/stat_msg.hh:26) is of an enum type. There are no guarantees that an enum has the size of an int. It usually has, but there are no guarantees. The C++ standard only states that the underlying type <i>is no larger than</i> int. Since C++11 you can force an underlying type to be used, but that does not help in this case. The declaration of the enum in libnml/rcs/rcs.hh:31 also means that it is <i>not</i> a scoped enumeration. Otherwise it would have been int by default, but the use-case here cannot make such guarantees. Therefore, there are possible size and alignment issues in the cast to int&.

One solution would be to declare RCS_STAT_MSG::status as int and the type-punning removed. However, there are many instances that then would require int casts in comparisons. Therefore, leaving it as is would be preferable. Removing the type-pun can be accomplished by using a temporary as shown in the diff below. It simply copies the current status member value into a local int, calls the CMS::update() and copies the result back in the status member.
```diff
diff --git a/src/libnml/nml/stat_msg.cc b/src/libnml/nml/stat_msg.cc
index 7e1301c287..830a930d06 100644
--- a/src/libnml/nml/stat_msg.cc
+++ b/src/libnml/nml/stat_msg.cc
@@ -35,7 +35,9 @@ int RCS_STAT_MSG_format(NMLTYPE t, void *buf, CMS * cms)
 {
     cms->update(((RCS_STAT_MSG *) buf)->command_type);
     cms->update(((RCS_STAT_MSG *) buf)->echo_serial_number);
-    cms->update((int&)(((RCS_STAT_MSG *) buf)->status));
+    int i = (int)((RCS_STAT_MSG *) buf)->status;
+    cms->update(i);
+    ((RCS_STAT_MSG *) buf)->status = (RCS_STATUS)i;
     cms->update(((RCS_STAT_MSG *) buf)->state);
 
     switch (t) {
```

--- a/src/libnml/nml/stat_msg.cc
+++ b/src/libnml/nml/stat_msg.cc
@@ -35,7 +35,7 @@ int RCS_STAT_MSG_format(NMLTYPE t, void *buf, CMS * cms)
 {
     cms->update(((RCS_STAT_MSG *) buf)->command_type);
     cms->update(((RCS_STAT_MSG *) buf)->echo_serial_number);
-    cms->update((int&)(((RCS_STAT_MSG *) buf)->status));
+    cms->update(((RCS_STAT_MSG *) buf)->status_int);
     cms->update(((RCS_STAT_MSG *) buf)->state);
 
     switch (t) {

--- a/src/libnml/nml/stat_msg.hh
+++ b/src/libnml/nml/stat_msg.hh
@@ -23,7 +23,22 @@ class RCS_STAT_MSG:public NMLmsg {
     RCS_STAT_MSG(NMLTYPE t, size_t sz);
     NMLTYPE command_type;
     int echo_serial_number;
-    RCS_STATUS status;
+    // The anonymous union is provided to allow for RCS_STAT_MSG_format() in
+    // nml/stat_msg.cc to call cms->update() with an integer reference. Member
+    // 'status' is normally accessed. Casting the 'status' member to int& in
+    // the cms->update() call will give a 'type-punned pointer dereference'
+    // warning. We sidestep this problem with the union where we can address
+    // either field and get the same value.
+    // The union is necessary because the update() call implementation checks
+    // the address of its argument to be in a specific memory region, which
+    // excludes using a temporary. Now, with the union, both members 'status'
+    // and 'status_int' share the same memory location. The 'RCS_STATUS' enum
+    // in rcs/rcs.hh explicitly declares the enum's underlying type as int.
+    // Therefore, both union members are of same underlying type.
+    union {
+        RCS_STATUS status;
+        int status_int;
+    };
     int state;
 };
 

--- a/src/libnml/rcs/rcs.hh
+++ b/src/libnml/rcs/rcs.hh
@@ -28,7 +28,10 @@ class RCS_TIMER;
 class RCS_CMD_MSG;
 class RCS_STAT_MSG;
 
-enum class RCS_STATUS {               /* Originally from nml_mod.hh */
+// The underlying type specifier of int is meant to stress that it
+// is of paramount importance that the underlying type of the
+// RCS_STATUS enum is int. See nml/stat_msg.hh for more details.
+enum class RCS_STATUS : int {           /* Originally from nml_mod.hh */
     UNINITIALIZED = -1,
     DONE = 1,
     EXEC = 2,

